internal/unionstore/pipelined_memdb.go

```diff
@@ -0,0 +1,159 @@
+}
+
+// Dirty returns whether the pipelined buffer is mutated.
+func (p *PipelinedMemDB) Dirty() bool {
+	return p.memBuffer.Dirty() || p.flushingMemBuffer != nil
+}
+
+// Get the value by given key, it returns tikverr.ErrNotExist if not exist.
+// The priority of the value is MemBuffer > flushingMemBuffer MemBuffer > Snapshot.
+func (p *PipelinedMemDB) Get(k []byte) ([]byte, error) {
+	v, err := p.memBuffer.Get(k)
+	if p.flushingMemBuffer != nil && tikverr.IsErrNotFound(err) {
+		v, err = p.flushingMemBuffer.Get(k)
+	}
+	if err != nil {
+		return v, err
+	}
+	if len(v) == 0 {
+		return nil, tikverr.ErrNotExist
+	}
+	return v, nil
+}
+
+func (p *PipelinedMemDB) GetFlags(k []byte) (kv.KeyFlags, error) {
+	f, err := p.memBuffer.GetFlags(k)
+	if p.flushingMemBuffer != nil && tikverr.IsErrNotFound(err) {
+		f, err = p.flushingMemBuffer.GetFlags(k)
+	}
+	if err != nil {
+		return 0, err
+	}
+	return f, nil
+}
+
+// MayFlush is called during execution of a transaction, it does flush when there are enough keys and the ongoing flushingMemBuffer is done.
+func (p *PipelinedMemDB) MayFlush() error {
+	if p.flushFunc == nil {
+		return nil
+	}
+	size := p.memBuffer.Size()
+	if size < MinFlushSize || (p.memBuffer.Len() < MinFlushKeys && size < MaxFlushSize) {
+		return nil
+	}
+	if !p.onFlushing.CompareAndSwap(false, true) && size < MaxFlushSize {
+		return nil
+	}
+	if p.flushingMemBuffer != nil {
+		if err := <-p.errCh; err != nil {
+			return err
+		}
+		if size >= MaxFlushSize {
+			// if the flush is triggered by size too large, the CAS operation may fail, manually set onFlushing to true in this case.
+			p.onFlushing.Store(true)
+		}
+	}
+	p.flushingMemBuffer = p.memBuffer
+	p.len += p.flushingMemBuffer.Len()
+	p.size += p.flushingMemBuffer.Size()
+	p.memBuffer = newMemDB()
+	go func() {
+		metrics.TiKVPipelinedFlushLenHistogram.Observe(float64(p.flushingMemBuffer.Len()))
+		metrics.TiKVPipelinedFlushSizeHistogram.Observe(float64(p.flushingMemBuffer.Size()))
+		err := p.flushFunc(p.flushingMemBuffer)
+		p.onFlushing.Store(false)
+		// Send the error to errCh after onFlushing status is set to false.
+		// this guarantees the onFlushing.Store(true) in another goroutine's MayFlush happens after onFlushing.Store(false) in this function.
+		p.errCh <- err
+	}()
+	return nil
+}
+
+// Flush will force flushingMemBuffer all the mutations, if there is an ongoing flush memdb, it will wait.
+// this should be called before running commit protocol.
+func (p *PipelinedMemDB) Flush() error {
+	// wait until the ongoing flush is done.
+	if p.flushingMemBuffer != nil {
+		if err := <-p.errCh; err != nil {
+			return err
+		}
+	}
+	if p.memBuffer.Len() > 0 {
+		metrics.TiKVPipelinedFlushLenHistogram.Observe(float64(p.memBuffer.Len()))
+		metrics.TiKVPipelinedFlushSizeHistogram.Observe(float64(p.memBuffer.Size()))
+		return p.flushFunc(p.memBuffer)
+	}
+	return nil
+}
+
+// GetMemDB returns the mutable memdb for writing.
+func (p *PipelinedMemDB) GetMemDB() *MemDB {
```

Comment:
Or is it needed to abstract the interfaces for `Memory DB` and just provide extrenal interface implementations, so the users may call `Set` or `Get` directly instead of exposing the internal `MemDB`. Maybe we would change the implementation doing further optimizations.